File: teaching/taltech2019-dev/code/talk/thursday/Ackermann.fs
#light "off"
module Ackermann
open Prims
open FStar.Pervasives

let rec ackermann : Prims.nat  ->  Prims.nat  ->  Prims.nat = (fun ( m  :  Prims.nat ) ( n  :  Prims.nat ) -> (match ((Prims.op_Equality m (Prims.parse_int "0"))) with
| true -> begin
(n + (Prims.parse_int "1"))
end
| uu____41 -> begin
(match ((Prims.op_Equality n (Prims.parse_int "0"))) with
| true -> begin
(ackermann (m - (Prims.parse_int "1")) (Prims.parse_int "1"))
end
| uu____58 -> begin
(ackermann (m - (Prims.parse_int "1")) (ackermann m (n - (Prims.parse_int "1"))))
end)
end))


let rec swapped_ackermann : Prims.nat  ->  Prims.nat  ->  Prims.nat = (fun ( n  :  Prims.nat ) ( m  :  Prims.nat ) -> (match ((Prims.op_Equality m (Prims.parse_int "0"))) with
| true -> begin
(n + (Prims.parse_int "1"))
end
| uu____121 -> begin
(match ((Prims.op_Equality n (Prims.parse_int "0"))) with
| true -> begin
(swapped_ackermann (Prims.parse_int "1") (m - (Prims.parse_int "1")))
end
| uu____138 -> begin
(swapped_ackermann (swapped_ackermann (n - (Prims.parse_int "1")) m) (m - (Prims.parse_int "1")))
end)
end))

File: teaching/taltech2020/code/talk/ocaml-out/Ackermann.ml
open Prims
let rec (ackermann : Prims.nat -> Prims.nat -> Prims.nat) =
  fun m ->
    fun n ->
      if m = Prims.int_zero
      then n + Prims.int_one
      else
        if n = Prims.int_zero
        then ackermann (m - Prims.int_one) Prims.int_one
        else ackermann (m - Prims.int_one) (ackermann m (n - Prims.int_one))
let rec (swapped_ackermann : Prims.nat -> Prims.nat -> Prims.nat) =
  fun n ->
    fun m ->
      if m = Prims.int_zero
      then n + Prims.int_one
      else
        if n = Prims.int_zero
        then swapped_ackermann Prims.int_one (m - Prims.int_one)
        else
          swapped_ackermann (swapped_ackermann (n - Prims.int_one) m)
            (m - Prims.int_one)
let (swapped_ackermann_bad : Prims.nat -> Prims.nat -> Prims.nat) =
  fun n -> fun m -> failwith "Not yet implemented:swapped_ackermann_bad"
